internal/pkg/sourceinfer/analyzer.go

```diff
@@ -0,0 +1,336 @@
+	for _, m := range builtSSA.Pkg.Members {
+		t, ok := m.(*ssa.Type)
+		if !ok {
+			continue
+		}
+		n, ok := t.Type().(*types.Named)
+		if !ok {
+			continue
+		}
+		s, ok := n.Underlying().(*types.Struct)
+		if !ok {
+			continue
+		}
+		for i := 0; i < s.NumFields(); i++ {
+			namedTypes := map[*types.Named]bool{}
+			findNamedTypes(s.Field(i).Type(), namedTypes)
+			for nt := range namedTypes {
+				obj := nt.Obj()
+				if len(typeGraph[obj]) == 0 {
+					typeGraph[obj] = make([]types.Object, 0, 1)
+				}
+				typeGraph[obj] = append(typeGraph[obj], n.Obj())
+			}
+		}
+	}
+}
+
+func findNamedTypes(t types.Type, namedTypes map[*types.Named]bool) {
+	deref := utils.Dereference(t)
+	switch tt := deref.(type) {
+	case *types.Named:
+		namedTypes[tt] = true
+	case *types.Array:
+		findNamedTypes(tt.Elem(), namedTypes)
+	case *types.Slice:
+		findNamedTypes(tt.Elem(), namedTypes)
+	case *types.Chan:
+		findNamedTypes(tt.Elem(), namedTypes)
+	case *types.Map:
+		findNamedTypes(tt.Elem(), namedTypes)
+		findNamedTypes(tt.Key(), namedTypes)
+	case *types.Basic, *types.Struct, *types.Tuple, *types.Interface, *types.Signature:
+		// these types cannot hold named types
+	case *types.Pointer:
+		// this should be unreachable due to the dereference above
+	default:
+		// The above should be exhaustive.  Reaching this default case is an error.
+		fmt.Printf("unexpected type received: %T %v; please report this issue\n", tt, tt)
+	}
+}
+
+func inferSources(pass *analysis.Pass, conf *config.Config, typeGraph ObjectGraph) ResultType {
+	inferredSources := ResultType{}
+
+	order := topoSort(typeGraph)
+
+	seen := map[types.Object]bool{}
+	for i := len(order) - 1; i >= 0; i-- {
+		o := order[i]
+		if seen[o] {
+			continue
+		}
+		if !(isSourceType(conf, o.Type()) || pass.ImportObjectFact(o, &inferredSource{})) {
+			continue
+		}
+		seen[o] = true
+		stack := []types.Object{o}
+		for len(stack) > 0 {
+			current := stack[len(stack)-1]
+			stack = stack[:len(stack)-1]
+			if current.Pkg() == pass.Pkg && current != o {
+				pass.ExportObjectFact(current, &inferredSource{})
+				inferredSources[current] = true
+			}
+			for _, n := range typeGraph[current] {
+				if seen[n] {
+					continue
+				}
+				seen[n] = true
+				stack = append(stack, n)
+			}
+		}
+	}
+
+	return inferredSources
+}
+
+// topoSort produces the topological order of a given graph.
+// If the graph contains cycles, the order within a cycle is not specified,
+// but the order of nodes outside the cycle will not be affected.
+func topoSort(graph ObjectGraph) []types.Object {
+	var order []types.Object
+	visited := map[types.Object]bool{}
+	var visit func(o types.Object)
+	visit = func(o types.Object) {
+		if visited[o] {
+			return
+		}
+		visited[o] = true
+		for _, n := range graph[o] {
+			visit(n)
+		}
+		order = append(order, o)
+	}
+	for o := range graph {
+		visit(o)
+	}
+	return order
+}
+
+func isSourceType(c *config.Config, t types.Type) bool {
```

Comment:
This is duplicated from `source.go`. It is my intent that the `source` analyzer will not need this function once it consumes this analyzer, either directly or via the `sourcetype` analyzer.

In the meantime, I could move it to `utils` if you think that would make more sense.

Reply:
Alternatively, this could easily be re-implemented in terms of `findNamedTypes` (defined above).

Reply:
I think I mentioned it offline, but if the boundary for this analyzer is to identify named source types, it likely should replace the current `sourcetypes` analyzer.  It makes sense, too, for it to occur early in the analyzer dependency chain, with `ast` cutting lower level than `ssa`.

That should be a future PR, because I would like to maintain the ability to see identified sources and fields that are currently presented by `sourcetypes`.  Having this analyzer absorb that feature can be part of the PR that replaces the `sourcetypes` analyzer.

Reply:
That makes sense to me!